### PR TITLE
pgp purge command

### PR DESCRIPTION
**Pull request by patrickxb**
Added `keybase pgp purge` command.

Every time it runs, it exports any pgp keys in lksec to saltpack encrypted files in $HOME.

If run with `-p` flag, it will also remove the pgp keys from lksec.

Idea was that they would run `keybase pgp purge`, make sure they could store the keys somewhere else, then run `keybase pgp purge -p` to actually remove them.

r? @maxtaco

### CODE REVIEW

File: go/client/cmd_pgp_purge.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+func NewCmdPGPPurge(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
+	return cli.Command{
+		Name:  "purge",
+		Usage: "Purge all PGP keys from Keybase keyring",
+		Action: func(c *cli.Context) {
+			cl.ChooseCommand(&CmdPGPPurge{Contextified: libkb.NewContextified(g)}, "purge", c)
+		},
+		Flags: []cli.Flag{
+			cli.BoolFlag{
```
Comment (maxtaco) [inline go/client/cmd_pgp_purge.go]: allow writing to a particular file?

File: go/client/cmd_pgp_purge.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	dui := s.G().UI.GetDumbOutputUI()
+	if len(res.Filenames) == 0 {
+		dui.Printf("No PGP keys found in local keyring\n")
+		return nil
+	}
+
+	dui.Printf("Exported PGP key files:\n")
+	for i, name := range res.Filenames {
```
Comment (maxtaco) [inline go/client/cmd_pgp_purge.go]: why might there be more than one?  one key per file?

File: go/engine/pgp_purge.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+// Run starts the engine.
+func (e *PGPPurge) Run(ctx *Context) error {
+	me, err := libkb.LoadMe(libkb.NewLoadUserPubOptionalArg(e.G()))
+	if err != nil {
+		return err
+	}
+	e.me = me
+
+	for _, bundle := range e.me.GetActivePGPKeys(false) {
```
Comment (maxtaco) [inline go/engine/pgp_purge.go]: I worry we might miss a key that's not "active".  Is it possible to iterate over each row in the secrets.mpack file instead?

File: go/engine/pgp_purge.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+func (e *PGPPurge) export(ctx *Context, bundle *libkb.PGPKeyBundle) error {
+	skb, key, err := e.findAndUnlock(bundle)
+	if err != nil {
+		return err
+	}
+	if key == nil {
+		return nil
+	}
+
+	filename := filepath.Join(e.G().Env.GetHome(), key.GetFingerprint().String()+".sp")
```
Comment (maxtaco) [inline go/engine/pgp_purge.go]: i would vote they wind up in the config directory next to the secrets.mpack file.  As long as we tell the user where they wind up, I think we're OK...

Comment (maxtaco) [inline go/engine/pgp_purge.go]: @oconnor663 do we have a suffix that we use for saltpacked files?

File: go/engine/pgp_purge.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+		return nil, nil, err
+	}
+	if skb == nil {
+		return nil, nil, nil
+	}
+
+	secretRetriever := libkb.NewSecretStore(e.G(), e.me.GetNormalizedName())
+	// the whole point of this is that these keys are unlockable without a prompt
+	// so this should suffice:
+	gk, err := skb.UnlockNoPrompt(nil, secretRetriever)
```
Comment (maxtaco) [inline go/engine/pgp_purge.go]: I definitely hear you, but do we need this assumption?  What if we allow prompting, does something bad happen?

File: go/engine/pgp_purge_test.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+	// redo, should purge 0 files
+
+	eng = NewPGPPurge(tc.G, keybase1.PGPPurgeArg{DoPurge: true})
+	if err := RunEngine(eng, ctx); err != nil {
+		t.Fatal(err)
+	}
+
+	if len(eng.KeyFiles()) != 0 {
+		t.Fatalf("number of exported key files: %d, expected 0", len(eng.KeyFiles()))
```
Comment (maxtaco) [inline go/engine/pgp_purge_test.go]: is it possible open to SKB file and ensure the purged key isn't in there?

File: go/engine/pgp_purge.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	var buf bytes.Buffer
+	if err := bundle.EncodeToStream(libkb.NopWriteCloser{W: &buf}, true); err != nil {
+		return err
+	}
+
+	arg := &SaltpackEncryptArg{
+		Source: &buf,
+		Sink:   out,
+	}
+	eng := NewSaltpackEncrypt(arg, e.G())
```
Comment (maxtaco) [inline go/engine/pgp_purge.go]: is the output armored?  If not, I recommend we armor it...

**Comment by maxtaco** [comment]
ok, made a pass through the code!

File: go/client/cmd_pgp_purge.go (hunk 2 above)
Comment (patrickxb) [inline go/client/cmd_pgp_purge.go]: yes, one key per file.

File: go/engine/pgp_purge.go (hunk 7 above)
Comment (patrickxb) [inline go/engine/pgp_purge.go]: armored

File: go/engine/pgp_purge.go (hunk 4 above)
Comment (patrickxb) [inline go/engine/pgp_purge.go]: Ok, i'll move them.

File: go/engine/pgp_purge.go (hunk 3 above)
Comment (patrickxb) [inline go/engine/pgp_purge.go]: I can change it, but before I do:  

1. What is the utility of exporting inactive PGP keys?  They are already unable to use them with keybase.  Wouldn't it be slightly confusing to get these keys again?
2. There can be multiple blocks for the same key.  The skb keyring uses last writer wins when making the index.  If iterating over each block, we could export multiple versions.  Should it iterate over the index instead?

File: go/engine/pgp_purge_test.go (hunk 6 above)
Comment (patrickxb) [inline go/engine/pgp_purge_test.go]: done

File: go/engine/pgp_purge.go (hunk 5 above)
Comment (patrickxb) [inline go/engine/pgp_purge.go]: I changed it.

**Comment by patrickxb** [comment]
@maxtaco exports/purges all blocks as we discussed

File: go/engine/pgp_purge.go [hunk 8]
```diff
@@ -61,80 +61,76 @@ func (e *PGPPurge) Run(ctx *Context) error {
...
-	if err != nil {
-		return nil, nil, err
-	}
-	pk, ok := gk.(*libkb.PGPKeyBundle)
-	if !ok {
-		return nil, nil, fmt.Errorf("unlocked key incorrect type")
+
+	for i, block := range blocks {
+		block.SetUID(e.me.GetUID())
+		key, err := block.PromptAndUnlock(promptArg, sstore, e.me)
```
Comment (maxtaco) [inline go/engine/pgp_purge.go]: if there are multiple keys, do you get prompted mulitple times, or just once?

File: go/engine/pgp_purge.go [hunk 9]
```diff
@@ -61,80 +61,76 @@ func (e *PGPPurge) Run(ctx *Context) error {
...
+		if err != nil {
+			return err
+		}
+
+		pgpKey, ok := key.(*libkb.PGPKeyBundle)
+		if !ok {
+			return fmt.Errorf("unlocked key incorrect type")
+		}
+
+		name := fmt.Sprintf("kb-%04d-%s.sp", i, pgpKey.GetFingerprint())
```
Comment (maxtaco) [inline go/engine/pgp_purge.go]: @oconnor663  any thoughts on what the saltpack suffix is?

**Comment by maxtaco** [comment]
One question, otherwise, LGTM!

File: go/engine/pgp_purge.go (hunk 8 above)
Comment (patrickxb) [inline go/engine/pgp_purge.go]: you shouldn't get prompted at all, but if you do, it should only happen once after secret store PR gets in.

Comment (maxtaco) [inline go/engine/pgp_purge.go]: ok !

File: go/engine/pgp_purge.go (hunk 9 above)
Comment (oconnor663) [inline go/engine/pgp_purge.go]: `.sp` seems reasonable to me. The only other think that comes to mind is straight up `.saltpack`, which feels too long.

Comment (maxtaco) [inline go/engine/pgp_purge.go]: Let's do `.saltpack`

**Comment by patrickxb** [comment]
local tests pass, merging.

File: go/engine/pgp_purge.go (hunk 9 above)
Comment (patrickxb) [inline go/engine/pgp_purge.go]: sorry, i missed this.  will change.

**Pull request merged**